### DISCUSSION TITLE
Yalmiptime is greatly increasing with iterations

### DISCUSSION THREAD

**Discussion by Coding-QQ** [yalmip/YALMIP#956] [Problems using YALMIP] [answered]
Hello:
        I'm using yalmip with some solvers like cplex, gurobi and intlinprog in MATLAB to solve MIP problems. In my code, there are many iterations, and in each of them, I will solve a similar MIP problem.  At first, one iteration only takes me about 10 seconds, But in the latter iterations, the running time is getting larger and larger. I checked the yalmiptime and solvertime in the last iteration, and found that the solver time is only 0.06s, but the yalmiptime is 20.39s!
       So how can I stop yalmiptime from getting longer? 

![pic](https://user-images.githubusercontent.com/68137987/127285485-f344098c-344f-4964-9d5c-ec991277c44e.png)
![pic2](https://user-images.githubusercontent.com/68137987/127286938-cc8680a5-37dc-4f47-a94d-7bcd120f57ea.png)

**Comment by johanlofberg** [comment, marked as answer]
Profilfoto för Johan Löfberg
Johan Löfberg<johanl@isy.liu.se>
12:48 (för 3 timmar sedan) 
till YALMIP
most likely you are creating new variables in the loop

for
define variables
define constraints
solve
end

you should try to move as much as possible outside the loop
define variables
define constraints
for
add stuff that changes
solve
end

or even better, use an optimizer approach instead

or at least use yalmip('clear') if you keep defining new stuff

**Reply by Coding-QQ** [reply]
Wow！I added yalmip('clear') before creating new variables in the loop and it worked!

for
yalmip('clear')
define variables
define constraints
solve
end

Thank you very much!

**Reply by SiaLight** [reply]
Hello:
    I meet the same problem,too.  But if I define some variables before the loop, this method won't work, because I'll use the variables I defined earlier in the loop.  The instruction 'yalmip('clear')' clarifies all variables, is there an instruction that clears some variables.
So how do I solve this problem?
   Thank you very much!

**Reply by EtONa-sys** [reply]
Hello:
I meet the same problem,too. And I want to know that how do I preallocate memory for constraints?
Thank you very much!

**Reply by johanlofberg** [reply]
The answer still applies.

Super bad
```
for loop
 define variables
 create constraints
 solve
end
```

Bad
```
for loop
 yalmip('clear')
 define variables
 create constraints
 solve
end
```

Less bad
```
define variables
for loop
 create constraints
 solve
end
```

Almost good
```
define variables
define fixed constraints
for loop
 combine changing constraints to fixed ones
 solve
end
```

Prefered
```
create optimizer object
for loop
 use optimizer object
end
```

**Reply by mrkento11** [reply]
Hello:
I am running an iterative optimization problem using both Yalmip and Gurobi. I call up all optimization within functions so my code looks like this:
```
Data analysis
define Parameters
while true loop
   matrix = function_Generation_Expansion_Problem_optimization(Parameters)
   verifier = function_UCP(matrix)
   if verifier == 1
       break
   else
      Parameters = Parameters + 1;
   end
end
```
On the other side, the functions Generation Expansion Problem and UCP are set up in this way:
```
yalmip('clear')
define variables
create constraints that depend on Parameters
solve
   matrix = value(optimized decision variable)
yalmip('clear')
end
```
By the hierarchy defined by you in this thread and in [THIS OTHER THREAD](https://groups.google.com/g/yalmip/c/po8p-419O14), I am clearly coding and calling up Yalmip/Gurobi in a BAD way. However, I think that it still should be working as I am clearing the yalmip variables, and MatLab should be tossing away all the variables defined in the functions. Nevertheless, I am running out of memory in a fairly rapid way.

Is there a way to make sure that yalmip is cleaning up all those variables? Perhaps calling the command yalmip('clear') before going into the function or using the [onCleanUp](https://www.mathworks.com/help/matlab/ref/oncleanup.html?s_tid=doc_ta) function?

Thanks for your time, Johan.

**Reply by johanlofberg** [reply]
If yalmip('clear') does not fix it, you've got something else going on

**Reply by Derrickzc** [reply]
define variables
 create constraints   (First Place)
for loop
 solve ==>has some result
 if (some result == true) 
  create "new" constraints; 
  (base on the First Place variable location)  
 end
end

Can yalmip  do this ?

**Reply by johanlofberg** [reply]
yes, constraints and variables are in no sense affected by the fact that you've called a solver. the only difference is that value now can return a value.

typical example of your loop
https://yalmip.github.io/The-cutsdp-solver

**Reply by mrkento11** [reply]
Hi again Dr. Löfberg,
Thanks for your help in previous questions
I have been trying to change my code as to add the 'optimizer' function. However, I am sometimes stumbling into unexpected errors when using the optimizer object.

My first question is: after creating the optimizer object, how/when should I use yalmip('clear')? Should I throw it every now and then so I don't run out of memory? How is the object affected by yalmip('clear')?

My second question is: I am having an inconsistency between the solution found between the optimizer object and the traditional way of setting up the yalmip problem. If I think of my optimization problem as min c*x, s.t. g(x) == 0; h(x) >= 0, then my optimization object becomes:  min c*x, s.t. g(x+a) == 0; h(x+a) >= 0. So the optimizer is a function only of a scalar a that affects directly a few constraints. So my question is if you think the inconsistency might be because the optimizer does not affect directly also the objective function c*x.

Thanks for your time.

**Reply by johanlofberg** [reply]
yalmip('clear') should never be necessary when using optimizer, as the whole idea is that you don't define any new variables or constraints, and thus nothing new is generated internally in YALMIP

No parameter in objective should not make any difference.

Impossible to say anything without an MRE
